etcd-recovery/etcdrecovery.go

```diff
@@ -0,0 +1,397 @@
+	_ = cmd.MarkFlagRequired("etcd-endpoints")
+
+	if len(opts.etcdEndpoints) != 3 {
+		panic(fmt.Errorf("etcd-endpoints must have 3 endpoints, got %d: %v", len(opts.etcdEndpoints), opts.etcdEndpoints))
+	}
+
+	cmd.AddCommand(NewStatusCommand(opts))
+	cmd.AddCommand(NewRecoverCommand(opts))
+
+	return cmd
+}
+
+func NewRecoverCommand(opts options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "recover",
+		Short:        "Recover etcd cluster",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGINT)
+			defer cancel()
+			err, failingEndpoint, failingPod, kubeclient := runStatus(ctx, opts)
+			if err != nil {
+				if failingEndpoint != nil {
+					fmt.Printf("failing etcd endpoint: %v\n", failingEndpoint)
+				}
+				if failingPod != nil {
+					fmt.Printf("failing etcd pod: %v\n", failingPod)
+				}
+				return runRecovery(ctx, *failingEndpoint, failingPod, kubeclient)
+			}
+
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func NewStatusCommand(opts options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "status",
+		Short:        "Determine if the HCP etcd cluster is healthy",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGINT)
+			defer cancel()
+
+			if err, failingEndpoint, failingPod, _ := runStatus(ctx, opts); err != nil {
+				if failingEndpoint != nil {
+					fmt.Printf("failing etcd endpoint: %v\n", failingEndpoint)
+				}
+				if failingPod != nil {
+					fmt.Printf("failing etcd pod: %v\n", failingPod)
+				}
+
+				return fmt.Errorf("ETCD cluster is unhealthy: %w", err)
+			}
+
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func runStatus(ctx context.Context, opts options) (error, *endpointStatus, *corev1.Pod, crclient.Client) {
+	fmt.Println("Checking etcd status")
+	// Generate kube client
+	kubeClient, err := getKubeClient()
+	if err != nil {
+		return fmt.Errorf("failed to get kube client: %w", err), nil, nil, nil
+	}
+
+	// Generate etcd client
+	etcdClient, err := getEtcdClient(opts)
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client: %w", err), nil, nil, kubeClient
+	}
+	defer etcdClient.Close()
+
+	err, failingEndpoint, failingPod := checkEtcdStatus(ctx, opts, etcdClient, kubeClient)
+	if err != nil {
+		return fmt.Errorf("failed to check etcd status: %w", err), nil, nil, kubeClient
+	}
+
+	if failingEndpoint != nil || failingPod != nil {
+		return fmt.Errorf("etcd cluster is unhealthy: %v", failingEndpoint), failingEndpoint, failingPod, kubeClient
+	}
+
+	fmt.Println("no etcd pods are failing, and no etcd endpoints are in error state, skipping recovery procedure")
+	return nil, nil, nil, nil
+}
+
+func runRecovery(ctx context.Context, failingEndpoint endpointStatus, failingPod *corev1.Pod, k crclient.Client) error {
```

Comment:
This is a completely transparent wrapper for `recoverEtcd`

Reply:
Make sense, let me fix it.